### PR TITLE
Add MeleeEnemy, HP regeneration

### PR DESCRIPTION
**Pull request by jeffsieu**
image
image

`MeleeEnemy` is an `Enemy` with a melee range. For now, it naively goes to the player by moving directly towards it.

When the player is within the range, it will initiate the attack sequence.
1. The enemy stops moving, and (presumably) charges the attack for a duration given by `preAttackPauseDuration`. The enemy color turns white. **This should be replaced by the attack animation.**
2. The enemy tries to attack the player immediately after the delay, dealing damage if the player is within range. The enemy color is changed back into red.
3. The enemy is then "paused" for another duration given by `postAttackPauseDuration`. This is to allow the player to execute melee hits more effectively. **This should be replaced by the post-attack animation.**

Health regeneration is implemented in `HealthEntity`. HealthEntities with a non-zero `RegenerationRate` will begin to regenerate their health after not being damaged for a duration given by `RegenerationDelay`, at a rate specified by `RegenerationRate`. Only the `Player` regenerates health for now.

A second progress bar is shown above the health bar, indicating the time left until health regeneration begins.

Enemies now also collide with one another, so they don't clump up.

TODO:
- Game over condition when player dies.

### CODE REVIEW

File: Scripts/Enemy/Enemy.cs [hunk 1]
```diff
@@ -23,16 +24,23 @@ public override void _Ready()
...
         MeshInstance cylinder = GetNode<MeshInstance>("MeshInstance");
         cylinder.SetSurfaceMaterial(0, material);
+
+        player = GetTree().Root.GetNode<Spatial>("Level").GetNode<Player>("Player");
+
+        RegenerationRate = 0.0f;
     }
 
-    public override void _PhysicsProcess(float delta)
+    public override void _Process(float delta)
```
Comment (owenl131) [inline Scripts/Enemy/Enemy.cs:33]: How come change from physics process to process?

Comment (jeffsieu) [reply Scripts/Enemy/Enemy.cs:33]: The UI stuff I think should put in Process, only physics-related stuff should put in PhysicsProcess I think. But so far all of our stuff we put in PhysicsProcess, which I'm not sure whether it's optimal or not.

**Comment by jeffsieu** [comment]
I think need to rethink how we do the HealthBars though, cos now the HealthEntity supposed to handle the updating of the HealthBars position in Process, but we should allow overriding of Process in subclasses of HealthEntity without forcing them to call base._Process().

**Comment by jeffsieu** [comment]
Maybe should implement HealthEntity as a "component" like Unity, then it just controls its parent

File: Scripts/Enemy/Enemy.cs (hunk 1 above)
Comment (owenl131) [reply Scripts/Enemy/Enemy.cs:33]: Oh I thought is use either one or the other

**Comment by owenl131** [comment]
> Maybe should implement HealthEntity as a "component" like Unity, then it just controls its parent

Agree

File: Scripts/Enemy/Enemy.cs (hunk 1 above)
Comment (owenl131) [reply Scripts/Enemy/Enemy.cs:33]: Ok but now its a bit funny cos the function is Process and it calls Base.PhysicsProcess

Comment (jeffsieu) [reply Scripts/Enemy/Enemy.cs:33]: got meh, i tot it's removed

Comment (owenl131) [reply Scripts/Enemy/Enemy.cs:33]: Gg see wrongly

**Comment by owenl131** [comment]
LGTM

**Review by daniellimws — CHANGES_REQUESTED** [review]
Turn off collider after enemy dies. And discuss about code for enemy behaviour in chat first. Otherwise code looks good.

File: Scripts/StateManager.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    public override void _Ready()
+    {
+        stateElapsedDelta = 0;
+        isPaused = false;
+    }
+
+    public override void _PhysicsProcess(float delta)
+    {
+        if (currentState != null && !isPaused)
```
Comment (daniellimws) [inline Scripts/StateManager.cs]: Can use `IsRunning` instead?

File: Scripts/StateManager.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public void GoTo(State state)
+    {
+        currentState = state;
+        stateElapsedDelta = 0;
+        isPaused = false;
+    }
+
+    public void Pause()
```
Comment (daniellimws) [inline Scripts/StateManager.cs:41]: Pause liao how to unpause

File: Scripts/Enemy/MeleeEnemy.cs [hunk 4]
```diff
@@ -2,65 +2,36 @@
...
-                hasAttacked = false;
-                attackDelta = 0;
-            }
+        // Let StateManager take over
+        if (stateManager.IsRunning)
             return;
-        }
-
 
         if (GlobalTransform.origin.DistanceTo(player.GlobalTransform.origin) < range)
```
Comment (daniellimws) [inline Scripts/Enemy/MeleeEnemy.cs:32]: Can this be in `Idle()`? Also let `stateManager` handle?

**Review by daniellimws — CHANGES_REQUESTED** [review]

**Review by daniellimws — APPROVED** [review]
Nicesah

**Review by daniellimws — APPROVED** [review]

**Pull request merged**